include/rmm/mr/device/block_aligned_resource_adaptor.hpp

```diff
@@ -0,0 +1,172 @@
+ * to satisfy allocation requests. This adaptor wraps allocations and deallocations from Upstream
+ * with block size alignment.
+ *
+ * @tparam Upstream Type of the upstream resource used for allocation/deallocation.
+ */
+template <typename Upstream>
+class block_aligned_resource_adaptor final : public device_memory_resource {
+ public:
+  /**
+   * @brief Construct a new block size aligned resource adaptor using `upstream` to satisfy
+   * allocation requests.
+   *
+   * If the allocation size is smaller or equal to the block size, the default alignment from
+   * `upstream` is used; if the allocation size is larger than the block size, it is aligned up to
+   * the block size.
+   *
+   * @throws `rmm::logic_error` if `upstream == nullptr`
+   *
+   * @param upstream The resource used for allocating/deallocating device memory.
+   */
+  block_aligned_resource_adaptor(Upstream* upstream,
+                                 std::optional<std::size_t> block_size = std::nullopt)
+    : upstream_{upstream}, block_size_{block_size.value_or(default_block_size)}
+  {
+    RMM_EXPECTS(nullptr != upstream, "Unexpected null upstream resource pointer.");
+  }
+
+  block_aligned_resource_adaptor()                                      = delete;
+  ~block_aligned_resource_adaptor() override                            = default;
+  block_aligned_resource_adaptor(block_aligned_resource_adaptor const&) = delete;
+  block_aligned_resource_adaptor(block_aligned_resource_adaptor&&)      = delete;
+  block_aligned_resource_adaptor& operator=(block_aligned_resource_adaptor const&) = delete;
+  block_aligned_resource_adaptor& operator=(block_aligned_resource_adaptor&&) = delete;
+
+  /**
+   * @brief Get the upstream memory resource.
+   *
+   * @return Upstream* pointer to a memory resource object.
+   */
+  Upstream* get_upstream() const noexcept { return upstream_; }
+
+  /**
+   * @copydoc rmm::mr::device_memory_resource::supports_streams()
+   */
+  bool supports_streams() const noexcept override { return upstream_->supports_streams(); }
+
+  /**
+   * @brief Query whether the resource supports the get_mem_info API.
+   *
+   * @return bool true if the upstream resource supports get_mem_info, false otherwise.
+   */
+  bool supports_get_mem_info() const noexcept override { return upstream_->supports_streams(); }
+
+ private:
+  static constexpr std::size_t default_block_size = 4096;
+
+  /**
+   * @brief Allocates memory of size at least `bytes` using the upstream
+   * resource with block size alignment.
+   *
+   * @throws `rmm::bad_alloc` if the requested allocation could not be fulfilled
+   * by the upstream resource.
+   *
+   * @param bytes The size, in bytes, of the allocation
+   * @param stream Stream on which to perform the allocation
+   * @return void* Pointer to the newly allocated memory
+   */
+  void* do_allocate(std::size_t bytes, cuda_stream_view stream) override
+  {
+    return upstream_->allocate(block_align(bytes), stream);
```

Comment:
This doesn't actually guarantee that the returned pointer is a multiple of 4096.

Reply:
I'm not sure if that's required for GDS. Asking them to clarify. How would we even do that? Even calling `cudaMalloc` wouldn't guarantee that, right? Do we have to allocate a larger buffer and offset from it?

Reply:
Right, cudaMalloc only guarantees alignment to 256KiB, and I think its implementation is 512KiB. So yeah, I guess you would allocate slightly larger, and then align to the nearest aligned offset...  Seems much easier to suballocate aligned than to upstream allocate aligned.

Reply:
You can see how I do the aligned allocate here for the host memory resources: https://github.com/rapidsai/rmm/blob/cab60bfcae9751d69efa148ae52bf9130704e51d/include/rmm/detail/aligned.hpp#L115-L117

The trick is that with host memory you can stash the base pointer in the extra bytes. You can't efficiently do that with device memory. 

One way to do this is to keep a `std::map<void*,void*>` to map returned, aligned pointer to it's original, base pointer returned from the upstream in order to later deallocate the same pointer.

Reply:
Ok, implemented this by allocating slightly more and then return the extra back to the upstream resource. Since this adds some overhead, also added a parameter that the user can specify so that only allocations bigger than the threshold get aligned.